Extensions/PlatformBehavior/tests/findHighestFloorAndMoveOnTop.spec.js

```diff
@@ -0,0 +1,428 @@
+        character.setPosition(300, -190);
+        checkMoveOn(behavior, platformBehavior, -10, 10);
+        expect(character.getY()).to.be(-200);
+      });
+
+      it('can detect an obstacle a bit too high to follow', function () {
+        character.setPosition(300, -189.9);
+        checkObstacle(behavior, platformBehavior, -10, 10);
+      });
+    });
+  });
+
+  [
+    {
+      description: '(An edge crossing from the left to the right)',
+      mask: collisionMasks.square,
+      position: [200, -100],
+    },
+    {
+      description: '(A vertex inside and edges crossing at the top)',
+      mask: collisionMasks.topTriangle,
+      position: [200, -100],
+    },
+    {
+      description: '(An edge crossing from the left to the top)',
+      mask: collisionMasks.topLeftTriangle,
+      position: [200, 0],
+    },
+    {
+      description: '(An edge crossing from the right to the top)',
+      mask: collisionMasks.topRightTriangle,
+      position: [200, 0],
+    },
+  ].forEach(({ description, mask, position }) => {
+    describe(description, function () {
+      const runtimeScene = makeTestRuntimeScene();
+      const character = addCharacter(runtimeScene);
+      const behavior = character.getBehavior('auto1');
+
+      const platform = addPlatform(runtimeScene, mask);
+      platform.setCustomWidthAndHeight(300, 300);
+      platform.setPosition(position[0], position[1]);
+      const platformBehavior = platform.getBehavior('Platform');
+
+      it('can detect an obstacle overlapping the top', function () {
+        character.setPosition(300, 199.9);
+        checkObstacle(behavior, platformBehavior, -10, 10);
+      });
+
+      it('can detect that the platform is away downward', function () {
+        character.setPosition(300, 200);
+        checkNoFloor(behavior, platformBehavior, -10, 10);
+      });
+    });
+  });
+
+  describe('(A platform with an hitbox under and another one above)', function () {
+    const runtimeScene = makeTestRuntimeScene();
+    const character = addCharacter(runtimeScene);
+    const behavior = character.getBehavior('auto1');
+
+    const platform = addPlatform(runtimeScene, collisionMasks.horizontalTunnel);
+    platform.setCustomWidthAndHeight(200, 200);
+    platform.setPosition(250, -250);
+    const platformBehavior = platform.getBehavior('Platform');
+
+    it('can detect the tunnel ceiling', function () {
+      character.setPosition(300, -210.1);
+      checkObstacle(behavior, platformBehavior, -10, 10);
+    });
+
+    // The character won't collide the ceiling anymore when he follows the floor.
+    // So, the ceiling should not be seen as an obstacle.
+    // This can happen if a tunnel is going down
+    // TODO probably something to fix.
+    it.skip('can detect a floor to follow down', function () {
+      character.setPosition(300, -210);
+      checkMoveOn(behavior, platformBehavior, -10, 10);
+      expect(character.getY()).to.be(-200);
+    });
```

Comment:
When the character is in a tunnel, it can't follow down a slope if the ceiling is near its head.
I'll do some tests to see how it works on master.

Reply:
Actually, in both master and this PR, the character is blocked by the tunnel ceiling when following down the slope. It's because of what we discussed previously about the obstacle search being horizontal only. So, this is not a blocking issue.